lib/rdf/transaction.rb

```diff
@@ -2,72 +2,110 @@ module RDF
   ##
   # An RDF transaction.
   #
-  # Transactions consist of a sequence of RDF statements to delete from and
-  # a sequence of RDF statements to insert into a given named graph.
+  # Transactions provide an ACID scope for queries and mutations.
   #
-  # Repository implementations may choose to sub-class this class
-  # to provide transactional support for repository updates, when
-  # accessed through {RDF::Repository#begin_transaction}.
+  # Repository implementations may provide support for transactional updates
+  # by providing an atomic implementation of {Mutable#apply_changeset} and 
+  # responding to `#supports?(:transactions)` with `true`.
+  # 
+  # We carefully distinguish between read-only and read/write transactions,
+  # in order to enable repository implementations to take out the
+  # appropriate locks for concurrency control. Transactions are read-only
+  # by default; mutability must be explicitly requested on construction in
+  # order to obtain a read/write transaction.
   #
-  # @example Executing a transaction against a repository
-  #   repository = ...
-  #   RDF::Transaction.execute(repository) do |tx|
+  # Individual repositories may make their own sets of guarantees within the 
+  # transaction's scope. In case repository implementations should be unable
+  # to provide full ACID guarantees for transactions, that must be clearly 
+  # indicated in their documentation. If update atomicity is not provided, 
+  # `#supports?(:transactions)` must respond `false`.
+  #
+  # @example Executing a read-only transaction
+  #   repository = RDF::Repository.new
+  #
+  #   RDF::Transaction.begin(repository) do |tx|
+  #     tx.query(predicate: RDF::Vocab::DOAP.developer) do |statement|
+  #       puts statement.inspect
+  #     end
+  #   end
+  #
+  # @example Executing a read/write transaction
+  #   repository = RDF::Repository.new
+  #
+  #   RDF::Transaction.begin(repository, mutable: true) do |tx|
   #     subject = RDF::URI("http://example.org/article")
   #     tx.delete [subject, RDF::RDFS.label, "Old title"]
   #     tx.insert [subject, RDF::RDFS.label, "New title"]
   #   end
+  # 
+  # The base class provides a full, buffered implementation depending on 
+  # `RDF::Changeset` and using `Changeset#apply`. Custom `Repositories`
+  # can implement a minimial write-atomic transactions by overriding
+  # `#apply_changeset`. Minimal snapshot isolation for queries can be 
+  # implemented by passing a `:snapshot` to `options` on initialization.
+  #
+  # For datastores that support Transactions natively, it is recommended 
+  # to implement a custom `Transaction` subclass, setting `@tx_class` to
+  # default to that class for the `Repository`.
   #
+  # @see RDF::Changeset
+  # @see RDF::Mutable#apply_changeset
   # @since 0.3.0
   class Transaction
     include RDF::Mutable
+    include RDF::Queryable
+
+    extend Forwardable
+
+    def_delegators :@snapshot, :query_pattern, :query_execute
```

Comment:
Queryable includes Enumerable, so it seems that it seems that it may be necessary. But, the reason for this may have more to do with default implementation, rather than an abstract need for queryable to also be enumerable. It would require some re-factoring of Queryable::Enumerator.